### PR TITLE
fix: Print dangling comments for `break|continue|debugger`

### LINKED ISSUE #12655: `break;` and `continue;` statements without braces followed by inline comment followed by `else` breaks prettier
**Prettier 2.6.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAdKACLGBmEBOGAFANwkCUGw622AljsTPgK5yUBG+cAhgNYkMAeiEYYACzoBnDJAC2chDADkMuAA9p8KGDgYA7nQA2RjF158MdGDVoY4RqXrwQi5ErawBfdCAA0IBAADjB00FLIoDz4+BD6AArRCBEoPEb6PACeEQFcPGB8cDAAyjyKADJ0UHDIOGlOufj5hSVB+VUA5sjMbAFwchxwACZDw+U8UB0sPB1wAGIEcjwwoZPIIDwsMBD+IOIwckYA6pLwUm26xcnWdABu1pnrYFI5IFVO+DDxTR1LtfVwAIAKyk6mKnSMcAAiiwIPB-o5ASA2vgPusODxBkZdkF8FUYEc6EMJMgABwABgCuIgTiOTSC61xcA+txqAQAjrD4N9gikNlIALTVYbDXbcTl0bjfGZ-JB1REBJxyOjdVhIqQQ6FcmpygEBGCYwnE8TIABM+qaxk6AGEIAoeOtmQBWXYsJwAFUxKXlDRAtzYAEkoKNYMUwHiQgBBYPFGCZSEIpxeLxAA)

```sh
--parser babel
```

**Input:**

```jsx

for (;;) {
  if (true) break; // this comment's existence will break it
  else foo();
}


```

**Output:**

```jsx
_@https://prettier.io/lib/standalone.js:41:7587
b@https://prettier.io/lib/standalone.js:41:15205
D@https://prettier.io/lib/standalone.js:45:576
St/<@https://prettier.io/lib/standalone.js:115:7299
formatCode@https://prettier.io/worker.js:141:21
handleMessage@https://prettier.io/worker.js:78:36
self.onmessage@https://prettier.io/worker.js:36:14

```

**Expected behavior:**

Prettier should not break.

### PR DESCRIPTION
**Pull request by liuxingbaoyu**
## Description

Some comments after `break|continue|debugger` cannot be printed.

Fixes #12655

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
Unstable

**Prettier pr-13677**
[Playground link](https://deploy-preview-13677--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAzCAnABACgNx4CUWwAOlOVlgJaq4wYCucxARhnAIYDWWA9HywwAFtQDOWSAFspCGAHIJcAB7j4UMHCwB3agBs9Wdl17UY5PJSxw9Yregg5ClilAC+IADQgIABxjU0GLIoJwYGBDaAAphCMEonHranACewd7snGDccDAAypyyADLUUHDIqIl2GRhZOfm+WaUA5sgMzN5wUqxwACa9fUWcUM2MnM1wAGKYUpwwASPIIJyMMBBeIMIwUnoA6qLwYo2aeXFm1ABuZilLYGLpIKV2GDBRtc2zFVVw3gBWYso8i09HAAIqMCDwL62H4gRoYZ5LVicHp6Da+DClGC7ai9ETIAAcAAZvBiIHZdrVfEsMXBnhdyt4AI4Q+BvPzxZZiAC0ZT6fQ2HBZ1A4b3GnyQlRh3jsUmobSYsLEwLBrPKku+3hgKJxeOEyAATFravoWgBhCAyThLRJo7yMOwAFRR8Sl1RAF2YAEkoANYHkwJj-ABBX15GApEHQuxuNxAA)

```sh
--parser babel
```

**Input:**

```jsx
for (;;) {

  if (true) break // this comment's existence will break it
;
  else foo();

}
```

**Output:**

```jsx
for (;;) {
  if (true)
    break; // this comment's existence will break it
  else foo();
}

```

**Second Output:**

```jsx
for (;;) {
  if (true) break; // this comment's existence will break it
  else foo();
}

```

**Comment by liuxingbaoyu** [comment]
[repl](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzCAnABACgNx4CUWwAOlOVlgJaq4wYCucxARhnAIYDWWA9HywwAFtQDOWSAFspCGAHIJcAB7j4UMHCwB3agBs9Wdl17UY5PJSxw9Yregg5ClilAC+IADQgIABxjU0GLIoJwYGBDaAAphCMEonHranACewd7snGDccDAAypyyADLUUHDIqIl2GRhZOfm+WaUA5sgMzN5wUqxwACa9fUWcUM2MnM1wAGKYUpwwASPIIJyMMBBeIMIwUnoA6qLwYo2aeXFm1ABuZilLYGLpIKV2GDBRtc2zFVVw3gBWYso8i09HAAIqMCDwL62H4gRoYZ5LVicHp6Da+DClGC7ai9ETIAAcAAZvBiIHZdrVfEsMXBnhdyt4AI4Q+BvPzxZZiAC0ZT6fQ2HBZ1A4b3GnyQlRh3jsUmobSYsLEwLBrPKku+3hgKJxeOEyAATFravoWgBhCAyThLOkAVg2jDsABUUfEpdUQBdmABJKADWB5MCY-wAQT9eRgKRB0LsbjcQA)

Should have nothing to do with this PR, as is the current main branch.

**Review by fisker — APPROVED** [review]
Thank you!

File: tests/format/js/for/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -683,47 +683,35 @@ label1: for (;;) continue label1
...
-  continue;
-  // comment
+for (;;) continue;
+// comment
 
-for (;;)
-  break;
-  // comment
+for (;;) break;
+// comment
```
Comment (liuxingbaoyu) [inline tests/format/js/for/__snapshots__/jsfmt.spec.js.snap:690]: This is not a format change, but a pre-existing bug in the snapshot.
This aligns with the current main branch behavior.
https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzCAnABACklAZxi1SwEsotgBfASiwCMM4BDAawG4AdKAel6yQAtkIQwQAGhAQADjDLQCyUCwwYIAdwAKqhEpQsANhpYBPJVKYswbODADKLUQBkKcZKiME4ljNdsOMtYUAObIMBgArj4gcEIMcAAmiUnOLFAhkSwhcABimEIsMPIZyCAskTAQkiAAFjBChgDqtWTwBEFgcPZ6bWQAbm2mZWAEFiAU3hgwWn4hhR5eMQBWBAAe9qGGcACKkRDwi4beUkEYU2UMLAmGNTIYFDBNZIkwtcgAHAAMp+reTX4ZGV7nApv13FIAI77eCzWT6coEAC0UDgSSSNWY0LIzFm2QWSE8xxi3iEZHCURJW12MPchKWUhg12er3eSAATIy-GRDKEAMIQEQsMqggCsNUi3gAKtd9ESTiB+tEAJJQFKwexgB5yACCavsMFM2yO3mo1CAA

**Comment by liuxingbaoyu** [comment]
I don't understand why only node18 fails.😕

**Comment by fisker** [comment]
> I don't understand why only node18 fails.😕

`FULL_TEST=1 yarn test ...`

**Comment by fisker** [comment]
Seems you fixed a  idempotence issue, remove this line https://github.com/prettier/prettier/blob/f38111fec6c35b513370832a84bdac8b5663763d/tests/config/format-test.js#L42

File: tests/config/format-test.js [hunk 2]
```diff
@@ -367,7 +366,14 @@ async function runTest({
...
       // To keep eye on failed tests, this assert never supposed to pass,
       // if it fails, just remove the file from `unstableTests`
-      expect(secondOutput).not.toEqual(firstOutput);
+      try {
+        expect(secondOutput).not.toEqual(firstOutput);
+      } catch {
+        /** @see unstableTests */
+        throw new Error(
+          `There is an unstable test that is now becoming stable, please remove it in \`unstableTests\`.\n${filename}`
+        );
```
Comment (fisker) [inline tests/config/format-test.js:375]: 👍

**Comment by fisker** [comment]
Can you add a changelog?

**Comment by liuxingbaoyu** [comment]
Does fixing idempotency need to be stated in the changelog? That seems like the title is a bit too long.

**Comment by fisker** [comment]
You can use two changelog entry, name `13677.md` and `13677-2.md`

File: changelog_unreleased/javascript/13677-2.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+// Prettier stable
+
+for (const f in {})
+  break; // comment
+
+// Prettier main
+
+for (const f in {}) break; // comment
+```
```
Comment (fisker) [inline changelog_unreleased/javascript/13677-2.md:20]:
````suggestion
<!-- prettier-ignore -->
```js
// Input
for (const f in {}) break// comment
;

// Prettier stable
for (const f in {})
  break; // comment

// Prettier stable (second format)
for (const f in {}) break; // comment

// Prettier main
for (const f in {}) break; // comment
```
````

File: changelog_unreleased/javascript/13677.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+for (;;) {
+  if (a) break; // comment
+  else foo();
+  if(a) continue; // comment
+  else foo();
+  if(a) debugger; // comment
+  else foo();
+}
+```
```
Comment (fisker) [inline changelog_unreleased/javascript/13677.md:32]:
````suggestion
<!-- prettier-ignore -->
```js
// Input
for (;;) {
  if (a) break; // comment
  else foo();
  if(a) continue; // comment
  else foo();
  if(a) debugger; // comment
  else foo();
}

// Prettier stable
Error

// Prettier main
for (;;) {
  if (a) break; // comment
  else foo();
  if(a) continue; // comment
  else foo();
  if(a) debugger; // comment
  else foo();
}
```
````

**Review by fisker — APPROVED** [review]
LGTM, except the changelog format.

**Comment by thorn0** [comment]
Something seems not right here. These comments should be attached as trailing to those statements, not as dangling. Did you try to fix it that way @liuxingbaoyu ?

**Comment by liuxingbaoyu** [comment]
No.
babel has it as a trailing comment, but espree doesn't.
https://astexplorer.net/#/gist/134357615990b3d06424d14bf13c2831/07bcca7af325512345159821b3e8b0e7e937bc64
And I tested in prettier, same result even with babel, so I'm not sure if this is intentional, I didn't touch them.

**Comment by fisker** [comment]
He is talking about our own comment attach logic https://github.com/prettier/prettier/blob/next/src/main/comments.js

**Comment by liuxingbaoyu** [comment]
I still don't understand the comment logic here, but I can only do so much. 😅

**Comment by thorn0** [comment]
@liuxingbaoyu You correctly identified the problematic place in the code. But it needs a fix for a more general case as that logic (added in #12017) can add dangling comments to _any_ kind of statement (not only `break/continue/debugger`) whereas only few statements support them. The idea with adding dangling comments with a marker is valid, but they should've been added to `IfStatement`.

**Comment by liuxingbaoyu** [comment]
I don't understand what needs to be done, what is this comment expected as? Dangling Comments of `ifStatement`? But this behavior appears to be intentional in https://github.com/prettier/prettier/pull/12017.
Also it looks like the way in https://github.com/prettier/prettier/pull/12017/files#diff-0aaa38223318fb8c3d0133b3b8d8f4cbc48d37209f26f3b818398d1ccc28994aR184-R217 is similar to my original commit.

**Comment by thorn0** [comment]
Sorry, I don't have a full solution in mind. I'm just pointing out that the bug isn't specific to `break/continue/debugger` (examples: [one](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAZgAgBTohASkzwCcdVMBeTAJgG5MKA+TABgdQFpOjIoBnCABs4AOiEQA5tlQEGAennF8AHShwh-OJgBGAQxLY5akABoQEAA4xU0fslAGSEAO4AFAwnso9Ql3oBPe3MdEj0wAGs4GABlPQBbOAAZVHVkdF8tELDI6JjLcNTJZBgSAFc4czh4nTgAEzr6pL0oSTK9STgAMQgSeL0YG1bkED0ymAgzEAALGHihAHVp1Hh+ArA4GK8V1AA3FYCRsH5gtAE4Ehg3MMl+9MzKkAArfgAPGKKRAEUyiHh7zSPAokLQkEb6WpCKaWEipGALVB1GDTZAADlY5hhEC0CzClhGMLgoN2cCmAEdfvBrlZvKN+Jx1PV6lMSHAKahWdcOnckBlAeYtPFUCVyo9+J84D8-qTeQ9zDA9DoEUiUUgaPKwqghEUAMIQeI8kBEgCsUzKWgAKorvHysiBdhUAJJQRqwGJgWHWACCLpiMACIgBWgAvsGgA), [two](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAZgAgBTohASkwBMJNhNIoBnCAGzgDo6IBzbVIgX0wHcALVAw4BaEUQD0EzHggAdKHDrU4mAEYBDAE7YCAbhAAaEBAAOMVNGrJQ2rRF4AFbQmsoNdXhoCe142q0NMABrOBgAZQ0AWzgAGVRFZHQPFX9AkLDw0yCE1mQYLQBXOGM4KLU4YmJK2I0oVkKNVjgAMQgtKI0YC3rkEA1CmAgjEH4YKLoAdUF4amywOHDXVAsANxXvPrBqPzQaOC0YR0DWTqSUkpAAK2oAD3DchgBFQoh4c+VL7K0VLT7NCp0EamLQJGCTVDEGD8ZAADgADMYQRAVJNAqY+iC4L9VnARgBHV7wY5mNz9agiRSVSojLRwQmoOnHJpnJDJT7GFRRVD5IqXaiPOAvN54tkXYwwDRqCFQmFIABMEsCQlyAGEIFFWSBsQBWEaFFQAFSlbnZqRAq2KAEkoNVYOEwKDzABBW3hGDeBgfFRcLhAA)). I'll look into it if I have time.

File: changelog_unreleased/javascript/13677-2.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#### Fix Idempotence for `break|continue|debugger` (#13677 by @liuxingbaoyu)
```
Comment (fisker) [inline changelog_unreleased/javascript/13677-2.md:1]: Hi @liuxingbaoyu the dangling comment issue fixed by #15826, would you send this fix separately?

Comment (liuxingbaoyu) [reply changelog_unreleased/javascript/13677-2.md:1]: I tried it and the fix is no longer correct, it doesn't match the behavior of the main branch, so I won't open the new PR anymore.

Comment (fisker) [reply changelog_unreleased/javascript/13677-2.md:1]: Okay, thank you anyway.

**Pull request closed without merging**